[UnifiedPDF] Cursor should update correctly over text and annotations

aprotyas opened the pull request:
#### 686826bf82ebceae9f402c35356ee29e8b11778f
<pre>
[UnifiedPDF] Cursor should update correctly over text and annotations
<a href="https://bugs.webkit.org/show_bug.cgi?id=265866">https://bugs.webkit.org/show_bug.cgi?id=265866</a>
<a href="https://rdar.apple.com/118550951">rdar://118550951</a>

Reviewed by Tim Horton.

UnifiedPDFPlugin has the plumbing in place to send cursor updates to the
UI process, but does not actually know what annotations/text (or other
PDF elements of interest) exist under a given point in a PDF page. This
patch builds up our hit-testing logic to imbibe UnifiedPDFPlugin with
this knowledge.

The logic to build up our set of PDF elements under a point is fairly
straightforward as we consult (and map) the type of the annotation at a
given point reported by PDFKit. As for text and images, which are not
annotations, we have to call into CoreGraphics&apos; PDFPageLayout SPI that
returns a bitset indicating the presence of text and images under a
given page point. Note that we only do so for tagged PDFs, since the SPI
in question does not return any information for PDFs that are not
tagged.

* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:

Provide declarations required to call into `CGPDFDocumentIsTaggedPDF`
and `CGPDFPageLayoutGetAreaOfInterestAtPoint`.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h: Renamed from Source/WebKit/Platform/spi/ios/PDFKitSPI.h.

Move to a Cocoa subdirectory since this SPI declaration header is now
required in UnifiedPDFPlugin.mm.

Also, provide a declaration to be able to obtain a PDFPage object&apos;s
corresponding CGPDFPageLayout instance.

* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.mm:

Soft-link more PDFAnnotation-adjacent classes used for type equality
checks in our hit-testing logic.

Also, drive-by fix that sorts the PDFKit symbols alphabetically.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::toWebCoreCursorType):

Some drive-by ergonomics nicety to avoid having to type out the
`UniversalPDFPlugin::` prefix.

(WebKit::UnifiedPDFPlugin::isTaggedPDF const):

Add a method to query whether the PDFDocument is tagged.

(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):

The bulk of the logic described in the commit message is implemented
here.

* Tools/TestWebKitAPI/Configurations/Base.xcconfig:

Since PDFKitSPI.h was moved to a Cocoa subdirectory, include said path
in the cocoatouch header search path.

Canonical link: <a href="https://commits.webkit.org/271592@main">https://commits.webkit.org/271592@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35197c701351da7a3be5b8a043f3b0cc2b4b2b6c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28878 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7518 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30239 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31494 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26332 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29450 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9647 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4869 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26374 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29150 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6220 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24801 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5416 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5554 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25814 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32831 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26414 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26254 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31795 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5532 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3699 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29575 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7162 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6909 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6009 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6053 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/Shared/Cocoa/PDFKitSoftLink.h:
```diff
@@ -35,10 +35,13 @@ SOFT_LINK_FRAMEWORK_FOR_HEADER(WebKit, PDFKit)
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, PDFAnnotationChoiceWidget)
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, PDFAnnotationPopup)
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, PDFAnnotationText)
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, PDFAnnotationButtonWidget)
```
I claimed alphabetic ordering in the commit message and then I see this. 🙅 Will move around

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/71dc5eecfc468d07b8c180342f01b32041848d68

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
```diff
@@ -74,6 +74,8 @@ class PDFDocumentLayout {
     void updateLayout(WebCore::IntSize pluginSize);
     WebCore::FloatSize scaledContentsSize() const;
 
+    bool isTaggedPDF() const;
```
does this belong here? this object is really about the layout of pages into document space

aprotyas replied:
Good question, I kept it here for the lack of wanting to expose the `PDFDocument` object.

aprotyas replied:
Clearly `PDFDocumentLayout` needn't expose that still, because I can just access it through `PDFPluginBase`, so I'll move this.

aprotyas replied:
Introduced `PDFPluginBase::isTaggedPDF()` instead -- https://github.com/WebKit/WebKit/commit/71dc5eecfc468d07b8c180342f01b32041848d68

aprotyas committed:
[UnifiedPDF] Cursor should update correctly over text and annotations
https://bugs.webkit.org/show_bug.cgi?id=265866
rdar://118550951

Reviewed by Tim Horton.

UnifiedPDFPlugin has the plumbing in place to send cursor updates to the
UI process, but does not actually know what annotations/text (or other
PDF elements of interest) exist under a given point in a PDF page. This
patch builds up our hit-testing logic to imbibe UnifiedPDFPlugin with
this knowledge.

The logic to build up our set of PDF elements under a point is fairly
straightforward as we consult (and map) the type of the annotation at a
given point reported by PDFKit. As for text and images, which are not
annotations, we have to call into CoreGraphics' PDFPageLayout SPI that
returns a bitset indicating the presence of text and images under a
given page point. Note that we only do so for tagged PDFs, since the SPI
in question does not return any information for PDFs that are not
tagged.

* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:

Provide declarations required to call into `CGPDFDocumentIsTaggedPDF`
and `CGPDFPageLayoutGetAreaOfInterestAtPoint`.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h: Renamed from Source/WebKit/Platform/spi/ios/PDFKitSPI.h.

Move to a Cocoa subdirectory since this SPI declaration header is now
required in UnifiedPDFPlugin.mm.

Also, provide a declaration to be able to obtain a PDFPage object's
corresponding CGPDFPageLayout instance.

* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.mm:

Soft-link more PDFAnnotation-adjacent classes used for type equality
checks in our hit-testing logic.

Also, drive-by fix that sorts the PDFKit symbols alphabetically.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::toWebCoreCursorType):

Some drive-by ergonomics nicety to avoid having to type out the
`UniversalPDFPlugin::` prefix.

(WebKit::UnifiedPDFPlugin::isTaggedPDF const):

Add a method to query whether the PDFDocument is tagged.

(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):

The bulk of the logic described in the commit message is implemented
here.

* Tools/TestWebKitAPI/Configurations/Base.xcconfig:

Since PDFKitSPI.h was moved to a Cocoa subdirectory, include said path
in the cocoatouch header search path.

Canonical link: https://commits.webkit.org/271592@main
